code/dictation.py

```diff
@@ -1,102 +1,229 @@
+            if isinstance(item, grammar.vm.Phrase) else
+            item.split(" "))
+    return words
+
+no_space_before = set("\n .,!?;:-/%)]}\"")
+no_space_after = set("\n -/#@([{$£€¥₩₽₹\"")
+def needs_space_between(before: str, after: str) -> bool:
+    return (before != "" and after != ""
+            and before[-1] not in no_space_after
+            and after[0] not in no_space_before)
+
+def auto_capitalize(text, state = None):
+    """
+    Auto-capitalizes text. `state` argument means:
+
+    - None: Don't capitalize initial word.
+    - "sentence start": Capitalize initial word.
+    - "after newline": Don't capitalize initial word, but we're after a newline.
+      Used for double-newline detection.
+
+    Returns (capitalized text, updated state).
+    """
+    output = ""
+    # Imagine a metaphorical "capitalization charge" travelling through the
+    # string left-to-right.
+    charge = state == "sentence start"
+    newline = state == "after newline"
+    for c in text:
+        # Sentence endings & double newlines create a charge.
+        if c in ".!?" or (newline and c == "\n"):
+            charge = True
+        # Alphanumeric characters and commas/colons absorb charge & try to
+        # capitalize (for numbers & punctuation this does nothing, which is what
+        # we want).
+        elif charge and (c.isalnum() or c in ",:"):
+            charge = False
+            c = c.capitalize()
+        # Otherwise the charge just passes through.
+        output += c
+        newline = c == "\n"
+    return output, ("sentence start" if charge else
+                    "after newline" if newline else None)
+
+
+# ---------- DICTATION AUTO FORMATTING ---------- #
+class DictationFormat:
     def __init__(self):
         self.reset()
-        self.paused = False
-        ui.register("app_deactivate", lambda app: self.reset())
-        ui.register("win_focus", lambda win: self.reset())
-
-    def reset(self):
-        self.caps = True
-        self.space = False
-
-    def pause(self, paused):
-        self.paused = paused
-
-    def format(self, text):
-        if self.paused:
-            return text
-
-        result = ""
-        for word in text.split():
-            is_sentence_end = False
-            is_punctuation = False
-            if word in sentence_ends:
-                word = sentence_ends[word]
-                is_sentence_end = True
-
-            elif word in punctuation:
-                word = punctuation[word]
-                # do  nothing
-                is_punctuation = True
-
-            elif self.space:
-                result += " "
-
-            if self.caps:
-                word = word.capitalize()
-
-            result += word
-            self.space = "\n" not in word and word[-1] not in no_space_after_these
-            self.caps = is_sentence_end
 
-        return result
+    def reset(self, before=None):
+        self.before = ""
```

Comment:
Why is there a `before` parameter if it's not being used? Looks like there are no call sites that pass it in, so might as well remove the parameter.